keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,640 @@
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-gcepd
+provisioner: pd.csi.storage.gke.io
+parameters:
+  type: hyperdisk-balanced
+  provisioned-iops-on-create: '1000'
+  provisioned-throughput-on-create: '100Mi'
+volumeBindingMode: WaitForFirstConsumer
+```
+
+You can request iops and throughput similar to storage for a pvc:
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: example-pv-claim
+spec:
+  storageClassName: csi-gcepd
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 2048Gi
+      iops: 1000
+      throughput: 50Mi
+```
+
+A new ControllerModifyVolume RPC will be added. The current spec requires ["If a volume corresponding to the specified volume ID is already larger than or equal to the target capacity of the expansion request, the plugin SHOULD reply 0 OK"](https://github.com/container-storage-interface/spec/blob/master/spec.md). The change here will be if the capacity change is for iops and throughput, we will not require that to be larger than the current configuration.
+
+#### ControllerModifyVolume
+
+A Controller plugin SHALL implement this RPC call. This RPC allows the CO to change key QoS parameters(iops and throughput) of a volume. 
+
+This operation MUST be idempotent. The requested  iops and  throughput values can be greater than or less than the current values.
+
+In physically-limited environments, updating performance parameters independently from capacity may result in underutilized resources. The SP MAY enforce limits on the allowed values. If the requested iops, throughput and capacity are not compatible with each other, the SP MUST return the OUT_OF_RANGE error code.
+
+The provider SHOULD provide independent performance fine tuning of read/write iops and read/write throughput defined in this API. Current cloud providers like AWS, GCP and Azure have already provided APIs to fine tune independent iops and throughput. An example is iops per GB. A provider may link these parameters but this is not encouraged due to the change cannot be counted for k8s quota and limit. Instead they should use capacity like it is done today.
+
+```protobuf
+// ControllerServer is the server API for Controller service.
+type ControllerServer interface {
+        ...
+        ControllerModifyVolume(context.Context, *ControllerModifyVolumeRequest) (*ControllerModifyVolumeResponse, error)
+        ...
+}
+
+message ModifyVolumeRequest {
+  // Contains identity information for the existing volume.
+  // This field is REQUIRED.
+  int64 volume_id = 1
+  // This field is OPTIONAL.This allows the CO to specify the
+  // iops of the volume to be updated to. The unit is IO per
+  // second per volume. 
+  // If specified it MUST always be honored.
+  // A value of 0 is equal to an unspecified field value.
+  int64 iops = 2
```

Comment:
What is the value of these being first-class fields instead of just having a catch all field?
map[string]string attributes

This would allow the plugin to have it's own fields it can react to.